File: app/src/main/java/org/tasks/dialogs/Linkify.java
package org.tasks.dialogs;

import static java.util.Arrays.asList;
import static org.tasks.Strings.isNullOrEmpty;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.style.URLSpan;
import android.view.View;
import android.widget.TextView;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import javax.inject.Inject;
import org.tasks.R;
import org.tasks.injection.ActivityContext;

public class Linkify {

  private final Context context;
  private final DialogBuilder dialogBuilder;

  @Inject
  public Linkify(@ActivityContext Context context, DialogBuilder dialogBuilder) {
    this.context = context;
    this.dialogBuilder = dialogBuilder;
  }

  public void linkify(TextView textView) {
    linkify(textView, () -> {});
  }

  public void linkify(TextView textView, Runnable onClick) {
    if (textView.length() == 0) {
      return;
    }

    android.text.util.Linkify.addLinks(textView, android.text.util.Linkify.ALL);

    textView.setOnClickListener(
        v -> {
          if (textView.getSelectionStart() == -1 && textView.getSelectionEnd() == -1) {
            onClick.run();
          }
        });

    URLSpan[] spans;
    Spannable spannable;
    CharSequence text = textView.getText();
    if (text instanceof SpannableStringBuilder || text instanceof SpannableString) {
      spannable = (Spannable) text;
    } else {
      return;
    }
    spans = spannable.getSpans(0, text.length(), URLSpan.class);
    for (URLSpan span : spans) {
      int start = spannable.getSpanStart(span);
      int end = spannable.getSpanEnd(span);

      spannable.removeSpan(span);
      spannable.setSpan(new ClickHandlingURLSpan(span.getURL(), onClick), start, end, 0);
    }
  }

  private class ClickHandlingURLSpan extends URLSpan {

    private final Runnable onEdit;

    ClickHandlingURLSpan(String url, Runnable onEdit) {
      super(url);
      this.onEdit = onEdit;
    }

    @Override
    public void onClick(View widget) {
      String title;
      String edit = context.getString(R.string.TAd_actionEditTask);
      String action;
      String url = getURL();
      Uri uri = Uri.parse(url);
      String scheme = uri.getScheme();
      if (isNullOrEmpty(scheme)) {
        scheme = "";
      }
      switch (scheme) {
        case "tel":
          title = uri.getEncodedSchemeSpecificPart();
          action = context.getString(R.string.action_call);
          break;
        case "mailto":
          title = uri.getEncodedSchemeSpecificPart();
          action = context.getString(R.string.action_open);
          break;
        case "geo":
          title = uri.getEncodedQuery().replaceFirst("q=", "");
          try {
            title = URLDecoder.decode(title, "utf-8");
          } catch (UnsupportedEncodingException ignored) {
          }
          action = context.getString(R.string.action_open);
          break;
        default:
          title = url;
          action = context.getString(R.string.action_open);
          break;
      }
      dialogBuilder
          .newDialog(title)
          .setItems(
              asList(action, edit),
              (dialogInterface, selected) -> {
                if (selected == 0) {
                  Intent intent = new Intent(Intent.ACTION_VIEW);
                  intent.setData(uri);
                  context.startActivity(intent);
                } else {
                  onEdit.run();
                }
              })
          .show();
    }
  }
}

File: app/src/main/java/org/tasks/dialogs/Linkify.kt
package org.tasks.dialogs

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.text.Spannable
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.style.URLSpan
import android.text.util.Linkify
import android.view.View
import android.widget.TextView
import androidx.core.net.toUri
import androidx.core.text.util.LinkifyCompat
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.findViewTreeLifecycleOwner
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.withStarted
import dagger.hilt.android.qualifiers.ActivityContext
import kotlinx.coroutines.launch
import org.tasks.R
import org.tasks.extensions.Context.safeStartActivity
import org.jetbrains.compose.resources.getString
import tasks.kmp.generated.resources.Res
import tasks.kmp.generated.resources.action_open
import timber.log.Timber
import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import javax.inject.Inject

class Linkify @Inject constructor(
    @ActivityContext private val context: Context,
    private val dialogBuilder: DialogBuilder
) {
    fun linkify(tv: TextView) {
        if (tv.length() == 0) {
            return
        }
        safeLinkify(tv)
        setMovementMethod(tv)
    }

    fun setMovementMethod(
        tv: TextView,
        linkClickHandler: ((url: String) -> Boolean) = { false },
        rowClickHandler: (() -> Unit) = {}
    ) {
        tv.setOnClickListener {
            if (!tv.hasSelection()) {
                rowClickHandler()
            }
        }
        val text = tv.text
        if (text is SpannableStringBuilder || text is SpannableString) {
            val spannable = text as Spannable
            val spans = spannable.getSpans(0, text.length, URLSpan::class.java)
            for (span in spans) {
                val start = spannable.getSpanStart(span)
                val end = spannable.getSpanEnd(span)
                spannable.removeSpan(span)
                spannable.setSpan(
                    ClickHandlingURLSpan(span.url, linkClickHandler, rowClickHandler),
                    start,
                    end,
                    0
                )

            }
        }
    }

    private suspend fun Uri.actionLabel(): String =
        if (scheme == "tel") {
            context.getString(R.string.action_call)
        } else {
            getString(Res.string.action_open)
        }

    private inner class ClickHandlingURLSpan(
        url: String?,
        private val linkClickHandler: ((String) -> Boolean),
        private val rowClickHandler: (() -> Unit),
    ) : URLSpan(url) {
        override fun onClick(widget: View) {
            if (linkClickHandler(url)) {
                return
            }
            val uri = url.toUri().takeUnless { it.scheme.isNullOrBlank() } ?: "https://$url".toUri()
            val title = when (uri.scheme) {
                "tel", "mailto" -> uri.encodedSchemeSpecificPart
                "geo" -> {
                    uri
                        .encodedQuery
                        ?.replaceFirst("q=".toRegex(), "")
                        ?.let {
                            try {
                                URLDecoder.decode(it, "utf-8")
                            } catch (ignored: UnsupportedEncodingException) {
                                it
                            }
                        }
                }
                else -> url
            }
            val owner = widget.findViewTreeLifecycleOwner() ?: context as? LifecycleOwner
            if (owner == null) {
                Timber.d("No lifecycle owner for $uri, opening it directly")
                context.safeStartActivity(Intent(Intent.ACTION_VIEW, uri))
                return
            }
            owner.lifecycleScope.launch {
                val actionLabel = uri.actionLabel()
                owner.withStarted {
                    dialogBuilder
                        .newDialog(title)
                        .setItems(
                            listOf(
                                actionLabel,
                                context.getString(R.string.TAd_actionEditTask),
                            )
                        ) { _, selected ->
                            if (selected == 0) {
                                context.startActivity(Intent(Intent.ACTION_VIEW, uri))
                            } else {
                                rowClickHandler()
                            }
                        }
                        .show()
                }
            }
        }
    }

    companion object {
        fun safeLinkify(textView: TextView?, mask: Int = Linkify.ALL) {
            try {
                LinkifyCompat.addLinks(textView!!, mask)
            } catch (e: UnsatisfiedLinkError) {
                Timber.e(e)
            }
        }
    }
}
